### DISCUSSION TITLE
How to provide default value for fun arg

### DISCUSSION THREAD

**Discussion by 93578237** [tweag/nickel#1642] [Q&A] [answered]
Without default values I explicitly pass `platform_` as null
```
let install_gh_release = fun name_ repo_ tag_ platform_ =>
    let action = {
      name = "Install %{name_}",
      uses = "jaxxstorm/action-install-gh-release@v1.10.0",
      with = {
        repo = repo_,
        tag = tag_,
        cache = "enable",
      }
    }
    in
    if platform_ != null then
      action & { with = { platform = platform_ } }
    else
      action
```

**Comment by ErinvanderVeen** [comment]
Hej @q0w, I think this question was directed at the Nickel team (this is the Topiary repository). I've pinged them for you and transferred the discussion to their repository!

**Reply by 93578237** [reply]
Oh, yes, sorry

**Comment by vkleen** [comment, marked as answer]
There aren't default values for functions per se in Nickel. Two options come to mind. First, you can do the Nix thing and take a record as the single argument. In this record, you can have default values:
```nickel
let install_gh_release = fun { name, repo, tag, platform ? "<default>" } => ...
```

The second, and the one that I would recommend, is not using a function to generate this data at all, but instead encode it as a recursively defined record with some fields left to be merged in. Fields that should end up in the final configuration can be marked as `not_exported` to hide them from serialization in simple cases. Sometimes it's more comfortable to have a dedicated `output` attribute. For example,

```nickel
let gh_release = {
  name,
  repo,
  tag,
  platform | default = null,
  output =
    let name' = name in
    {
      name = "Install %{name'}",
      uses = "jaxxstorm/action-install-gh-release@v1.10.0",
      with.repo = repo,
      with.tag = tag,
      with.cache = "enable",
    } & (if platform == null then {} else {
      with.platform = platform,
    }),
} in
gh_release & {
  name = "some name",
  repo = "some repo",
  tag = "some tag",
}
```
Ideally one would accumulate these kinds of release until the very top of the configuration and only then select the `output` field in them. That way, you can keep as much overridability as possible.

**Reply by 93578237** [reply]
Thanks!
